add missing is_* macros

eksperimental opened the pull request:
I found this the hard way. I was relying on the list of this page, to realized that we were missing some functions/macros that are allowed in guards.

I have added the word  "macros" now since there is more than just one (is_nil)

@josevalim @ericmj I have the feeling that we need a helper function in elixir that lists the functions that are allowed in guard.
how about having: `Kernel.allowed_in_guard?(module, function_name, arity)` and  `Kernel.list_allowed_in_guards/0`

I think relying on a an external website to know what can be used in guards,
Regardless creating the helper functions, I think we need to create the list that we have in this page, in the elixir documentaion itself.

eksperimental commented:
@josevalim please do not merge this, as I am adding another commit for review

josevalim commented:
I am :-1: because any module can add their own guards. I think right now I would leave the list as is but improve the wording at the top to say:

    Elixir imports and allows the following expressions in guards by default:

(Instead of mentioning the Erlang VM)

josevalim commented:
Thoughts?

eksperimental commented:
ok, I have submmited a new commit.
I reviewed the document, and I can see some information didn't make it here maybe because it's expressed in different ways that those functions can be used in a guards. So while doing a grep, some were not showing up.

Comments on some of the changes:
[x] I removed `!` because it is not supported (only `not`)
[x] added the arithmetic unary opoerators
[x] added `--`, @josevalim can you tell me if the "as long as the left side is a literal" comment apply for this operator. By the what is the definition of "a literal", a link to it would be appreciated to have on this page.
[x] added a note at the bottom, explaining that the shortcircuit versions of the boolean operators are not allowed in guards

eksperimental commented:
Then, maybe have a function by that lists the guard safe expressions by default that come with Elixir.
maybe define a property inside the function, something like @guard_safe
If you consider appropriate, maybe it would be better to move this talk to the "elixir" repo or the mailing list, since not as many people follow this repository.

Also, I think we need to rephrase this
> Elixir imports and allows the following expressions in guards by default:
because some of those modules are not imported (Bitwise, Integer and Record)

josevalim commented:
I like your addition with Bitwise. :+1:

I wouldn't go with your rephrase though. The main point here is that there can be more, **anywhere**. If we end-up listing all the possible additions from Elixir, then it defeats the purpose of telling that there may be more. Although using one as example, and Bitwise is a really good one, will be great.

lexmag commented on getting-started/case-cond-and-if.markdown:
```diff
@@ -98,6 +105,9 @@ The Erlang Virtual Machine (VM) only allows a limited set of expressions in guar
 Additionally, users may define their own guard functions, usually starting with "is_".
 
+Note that while boolean operators such as (`and`, `or`, `not`) are allowed in guards,
+their shortcircuit conterparts are not allowed in guards such as (`&&`, `||`, `!`).
```
It's not true to name `&&`, `||`, `!` shortcircuit counterparts.

eksperimental replied:
Yeah..I wanst too sure about that. How should we refer to them as, or should we simply list them.

josevalim replied:
"while the more general and short-circuiting operators `&&`, `||` and `!` are not."

eksperimental commented:
@josevalim suggestion updated
So, should I remove some of the additions to the list?

josevalim commented:
@eksperimental please remove the Integer and Record ones. :)

eksperimental commented:
@josevalim done. can you confirm this please?
>> added --, @josevalim can you tell me if the "as long as the left side is a literal" comment apply for this operator. By the what is the definition of "a literal", a link to it would be appreciated to have on this page.

josevalim commented on getting-started/case-cond-and-if.markdown:
```diff
@@ -54,14 +54,16 @@ The first clause above will only match when `x` is positive.
+* boolean operators (`and`, `or`, `not`)
+* arithmetic operations (`+`, `-`, `*`, `/`)
+* arithmetic unary operators (`+`, `-`)
+* `<>` and `++`, `--` as long as the left side is a literal
```
`--` is not allowed in guards. Given your question we should probably rewrite this as:

    `<>` and `++`, as long as the left side is a binary or a list respectively

eksperimental commented:
http://elixir-lang.org/docs/master/elixir/Kernel.html#--/2
 docs say it is.

Sorry I cannot test it now.

josevalim commented:
It isn't, for sure. I will update the docs for `--`.

eksperimental commented:
@josevalim: `++` is not working. Am i missing something?

```elixir
iex(10)> defmodule Foo do
...(10)>   def bar(string) when is_bitstring(string) and string <> "1" == "01", do: :binary
...(10)>   def bar(list) when is_list(list) and list ++ [1] == [0, 1], do: :binary
...(10)> end
iex:10: warning: redefining module Foo
** (CompileError) iex:12: illegal guard expression
    (stdlib) lists.erl:1337: :lists.foreach/2
    (stdlib) erl_eval.erl:669: :erl_eval.do_apply/6
    (elixir) src/elixir.erl:195: :elixir.erl_eval/3
    (elixir) src/elixir.erl:183: :elixir.eval_forms/4
       (iex) lib/iex/evaluator.ex:117: IEx.Evaluator.handle_eval/5
```

josevalim commented:
You are right. `++` and `<>` do not really work in guards. They do work inside patterns though.

eksperimental commented:
well, actually `<>` is working fine, just `++` and `--` are not
Should it be?
> `<>`, as long as the left side is a binary

josevalim commented:
`<>` would work with any value then?

eksperimental commented:
sorry, I'm not sure if I understand your question,
what do you mean by any value?

```elixir
defmodule Foo do
  def equal?(string1, string2) when string1 <> string2 == string2 <> string1, do: true
  def equal?(_, _), do: false
end

iex(34)> Foo.equal?("elixir", "elixir")
true
iex(35)> Foo.equal?("a", "aa")
true
iex(36)> Foo.equal?("aa", "a")
true
iex(37)> Foo.equal?("aa", "ab")
false
```

josevalim commented:
Exactly that. Both left and right side need to be binaries but that is because `<>` works on binaries. The left and right side do not need to be literal though.

eksperimental commented:
sorry, but I still don't know what a literal is :-|

josevalim commented:
A literal is a list, a number, a string *literally* in the code, not in a variable. But if you are not sure what a literal is, we should probably not use it, because others would be confused. What the previous version meant was that this will work: `[1, 2, 3] ++ rest` but this won't `some_list ++ rest`. The left side needs to be literally a list.

eksperimental commented:
great. I got you now. Thanks for the explanation.

I'm still wondering if this is the right behaviour?

```elixir
defmodule Foo do
  def bar(arg1, arg2) when arg1 <> arg2, do: 1
  def bar(arg1, arg2) when arg1 <> arg2 == arg2 <> arg1, do: 2
  def bar(_arg1, _arg2) when 1 <> 2, do: 3
end

iex(43)> Foo.bar("1", "1")
2
iex(44)> Foo.bar("", "")  
2
iex(45)> Foo.bar(0, 1)  
** (FunctionClauseError) no function clause matching in Foo.bar/2
    iex:43: Foo.bar(0, 1)
```

Why `Foo.bar("1", "1")` doesn't return with the first definition. doesn't `"1" <> "1"` evaluate to truthy?,
Even doing it with literals it never matches that def.
`def bar(_arg1, _arg2) when ("0" <> "0"), do: 0`

and `Foo.bar(0, 1)` doesn't throw an "argument error", as just a regular pattern matching would do.

are we in the presence of a bug?

josevalim commented:
>  doesn't "1" <> "1" evaluate to truthy?

As with `and`, `or` and `not`, guards are strictly boolean. I believe we even have a note in this section.

> and Foo.bar(0, 1) doesn't throw an "argument error", as just a regular pattern matching would do.

We also mention that errors in guard simply make the guard fail (they don't raise).

eksperimental commented:
all good then.

so.. what is the final wording for that line?

> * `<>`as long as both sides are binaries
or
> * the concatenation operator `<>`, as long as both sides are binaries
or
> * the concatenation operator `<>`

I'm in favour of the last one, since I don't think it is necessary to describe what values accept every function.

josevalim commented:
:+1: for the last one. You can call it: "the binary concatenation operator `<>`"

eksperimental commented:
we are ready to go

ericmj commented on getting-started/case-cond-and-if.markdown:
```diff
@@ -54,13 +54,15 @@ The first clause above will only match when `x` is positive.
+* arithmetic unary operators (`+`, `-`)
+* the binary concatenation operator `<>`
+* the `in` operator as long as the right side is range or a list
+* all operators in the`Bitwise` module (`bnot`, `~~~`, `band`, `&&&`, `bor`, `|||`, `bxor`, `^^^`, `bsl`, `<<<`, `bsr`, `>>>`)
```
The Bitwise operators shouldn't be in the section about Elixir imports, they are not imported.

josevalim replied:
I can amend this myself. :)
